### ISSUE TITLE
unable to get ProtobufCFieldDescriptor for oneof message

### ISSUE THREAD

**Issue by pbor** [protobuf-c/protobuf-c#204]
if I have a message with one of like

message Wrapper {
  oneof msg {
    MsgA msg_a = 1;
    MsgB msg_b = 2;
  }
}

How can I get the offset at which I can write my message A or B?

In other words, if I do

desc = protobuf_c_message_descriptor_get_field_by_name(descriptor, "msg_a");
I can get desc->offset

But this requires knowing that the field is called "msg_a".
Is there any way to obtain the same offset only knowing that the oneof is called "msg"?

Basically the use case I have is that I have many wrapper messages that contain "oneof msg", but the actual fields inside the oneof change. I would like to be able to write a generic function that gives me the "unwrapped" message

**Comment by pbor** [comment]
For now I have worked around this by using fields[0] since I can assume that my wrapper messages only contain the oneof. But this is pretty ugly

**Comment by lipnitsk** [comment]
I don't think we currently propagate the name of the oneof into `ProtobufCFieldDescriptor`, but there is some reserved space in there, so it should be feasible. Since this hasn't seen any activity in years, is there still a need for such a feature?

**Comment by pbor** [comment]
I think the need is still there, I have the hack mentioned in the previous comment in my code, but it would be good to have a proper solution

**Comment by lipnitsk** [comment]
@pbor sorry for taking so long on this, but can you see if a file-level option would work for your use case? I have added `use_oneof_field_name` to override the field name to use the oneof name instead. Take a look at the test case and the branch in #466 and let me know if you have some time!

**Comment by pbor** [comment]
Hey Ilya! Cool to see bursts of activity on the project :)

Do you mean:

```
option (pb_c_file).use_oneof_field_name = true;
```

It is not clear to me what are the side effects of that... would that mean that I cannot access by name the fields within the oneof?

In general, if I have to be honest, I would not add those to my project... we use the same proto files from a whole bunch of languages and adding implementation specific annotations is a slippery slope I would like to avoid...

**Comment by lipnitsk** [comment]
> would that mean that I cannot access by name the fields within the oneof?

Correct, only if you set `use_oneof_field_name` to true, all fields contained in a oneof will now carry the same name, so you would have to use an index to access a particular field. Putting the oneof name into a reserved field instead of replacing the field name would require protobuf-c changes, and potentially a new API to look up fields based on oneof name which seems sub-optimal.


> In general, if I have to be honest, I would not add those to my project

While we can revisit what this particular option controls, I would like to move in the direction of specifying compiler options as proto options in the proto file, rather than command line options to protoc-c. This is good from a couple of perspectives:
- it is self-documenting - all protoc-c settings are contained within the proto file itself
- follows standard precedent in other protobuf implementations (similar to `java_package` option upstream, options for nanopb, etc)

Given the above, if you have a proposal that would work better for you while still using a file-level (or potentially a oneof-level) option I'd like to hear it.

**Comment by pbor** [comment]
> I would like to move in the direction of specifying compiler options as proto options in the proto file, rather than command line options to protoc-c. This is good from a couple of perspectives:
> 

When I said "I would not add it to my project" I meant the general concept of having compiler options in the .proto file. With regard to *this* specific option for `oneof`, I am ok with the approach, at least for now.

While I see the reasons you listed, having compiler options in the `.proto` files seems an anti pattern to me. The whole reason we looked at protobuf in the first place was to have a high level description of our protocol that abstracted us from the implementation. We want this description to be independent from the implementation language, so that there may be many implementations.

As a matter of fact, requiring the options in the .proto files would be problematic also on a practical level: our proto files are in a separate repository and different implementations of the protocol pull it in and generate the code in the language they prefer.
Albeit in this case we control the protobuf-c implementation, if tomorrow a third party want to do a Go or Rust implementation and they needed similar options, I do not like the idea of messing with the shared proto files. Even worse if two implementations using the same language wanted to use different options.

IMHO compiler options belong in the build system of the project that is consuming the proto files.

**Comment by lipnitsk** [comment]
One problem is that it's more cumbersome to specify message-level
(pb_c_msg) options using cmd line. For file-level (pb_c_file) it may work,
and I'll look if there is an easy way to leverage existing changes and add
another way to set the same options via cmd line.

Thanks for your feedback.

**Comment by lipnitsk** [comment]
Would something like
https://jpa.kapsi.fi/nanopb/docs/reference.html#defining-the-options-in-a-.options-file
work for your use case?

**Comment by pbor** [comment]
Oh cool. Yes, I guess that would do the trick.

**Comment by pbor** [comment]
I mean, even better if I have an option to tell the compiler where to look for the option files instead of forcing the same dir.

**Comment by lipnitsk** [comment]
At some point I'll look at adding a way to set file-level options via cmd line. Ideally this function should be supported upstream, see https://github.com/protocolbuffers/protobuf/issues/8225.

**Comment by lipnitsk** [comment]
We should probably have a separate issue to track that request since it's not for any specific option. This particular issue will be fixed in the next release, but it is unfortunate that you won't be able to make use of it - oh well.

**Comment by pbor** [comment]
It is ok, as mentioned right now we have worked around it for a while and it works :)

Thanks a lot for all your work on protobuf-c!
